examples/llm/src/models/layers/attention.py

```diff
@@ -12,377 +13,343 @@
+            attn_bias = query.new_zeros(b_size, 1, 1, s_k)
 
-    @staticmethod
-    def mask_shape(*args, **kwargs):
-        return None
+        attn_bias = attn_bias.masked_fill(
+            ~key_padding_mask.view((b_size, 1, 1, s_k)), -float('inf'))
 
-    @staticmethod
-    def attn_mask_(*args, **kwargs):
-        return None
+    query = rearrange(query, 'b s (h d) -> b s h d', h=n_heads)
+    key = rearrange(key, 'b s (h d) -> b s h d', h=n_heads)
+    value = rearrange(value, 'b s (h d) -> b s h d', h=n_heads)
+
+    attn_output = flash_attn_triton.flash_attn_func(query, key, value,
+                                                    attn_bias, is_causal,
+                                                    softmax_scale)
 
-    @staticmethod
-    def generate_attn_mask(
-        attn_mask,
-        batch_size,
-        heads,
-        seq_len,
-        key_padding_mask=None,
-        alibi=False,
-        dtype=None,
-    ):
-        return attn_mask  # None
+    output = attn_output.view(*attn_output.shape[:2], -1)
+    return output, None
 
 
-class TritonFlashCausalAttention(nn.Module):
-    """Multi-headed self attention using triton FlashAttn kernel.
+class MultiheadAttention(nn.Module):
+    """Multi-head self attention.
 
-    This also includes bias for Alibi integration.
+    Using torch or triton attention implemetation enables user to also use
+    additive bias.
     """
 
     def __init__(self, cfg: DictConfig, device: Optional[str] = None):
         super().__init__()
-        try:
-            from examples.llm.src.models.layers.flash_attention import (  # type: ignore
-                FlashAttention, FlashMHA)
-        except ImportError as e:
-            raise e
-
-        assert cfg.attn_pdrop == 0, 'triton kernel does not support attn_dropout'
+        self.attn_impl = cfg.get('attn_impl')
 
         self.clip_qkv = cfg.get('attn_clip_qkv')
         self.attn_qk_ln = cfg.get('attn_qk_ln')
+
         self.d_model = cfg.d_model
         self.n_heads = cfg.n_heads
+        self.softmax_scale = cfg.get('softmax_scale')
+        if self.softmax_scale is None:
+            self.softmax_scale = 1 / math.sqrt(self.d_model / self.n_heads)
+        self.attn_dropout_p = cfg.get('attn_pdrop')
+
+        self.Wqkv = nn.Linear(self.d_model,
+                              3 * self.d_model,
+                              bias=True,
+                              device=device)
+        # for param init fn; enables shape based init of fused layers
+        fuse_splits = (cfg.d_model, 2 * cfg.d_model)
+        self.Wqkv._fused = (0, fuse_splits)  # type: ignore
+
+        if self.attn_qk_ln:
+            layernorm_class = nn.LayerNorm if not cfg.get(
+                'low_precision_layernorm', False) else LPLayerNorm
+            self.q_ln = layernorm_class(self.d_model, device=device)
+            self.k_ln = layernorm_class(self.d_model, device=device)
+
+        self.is_causal = True  # causal attn impl
```

Comment:
I suggest deleting this and moving `is_causal: bool` to the `MultiheadAttention.forward()` args, because for PrefixLM (needed soon both internal and customer facing), we would like to set this to `False`, right?

Reply:
I added `self.is_causal` in prep for `PrefixLM`. Wasn't sure where to put it, but I'll move it to `MultiheadAttention.forward()` as suggested